src/audio/google/google_rtc_audio_processing.c

```diff
@@ -563,125 +591,191 @@ static int google_rtc_audio_processing_reset(struct processing_module *mod)
+
+	size_t num_of_bytes_to_process;
 	int num_samples_remaining;
 	int num_frames_remaining;
 	int channel;
-	int frames;
 	int nmax;
-	int ret;
-	int i, j, n;
 
-	struct input_stream_buffer *ref_streamb, *mic_streamb;
-	struct output_stream_buffer *out_streamb;
-	struct audio_stream *ref_stream, *mic_stream, *out_stream;
+	struct sof_source *ref_stream, *src_stream;
+	struct sof_sink *dst_stream;
+
+	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
 
 	if (cd->reconfigure) {
 		ret = google_rtc_audio_processing_reconfigure(mod);
 		if (ret)
 			return ret;
 	}
 
-	ref_streamb = &input_buffers[cd->aec_reference_source];
-	ref_stream = ref_streamb->data;
-	ref = audio_stream_get_rptr(ref_stream);
-
-	num_aec_reference_frames = input_buffers[cd->aec_reference_source].size;
-	num_aec_reference_bytes = audio_stream_frame_bytes(ref_stream) * num_aec_reference_frames;
-
-	num_samples_remaining = num_aec_reference_frames * audio_stream_get_channels(ref_stream);
-	while (num_samples_remaining) {
-		nmax = audio_stream_samples_without_wrap_s16(ref_stream, ref);
-		n = MIN(num_samples_remaining, nmax);
-		for (i = 0; i < n; i += cd->num_aec_reference_channels) {
-			j = cd->num_aec_reference_channels * cd->aec_reference_frame_index;
-			for (channel = 0; channel < cd->num_aec_reference_channels; ++channel)
-				cd->aec_reference_buffer[j++] = ref[channel];
-
-			ref += audio_stream_get_channels(ref_stream);
-			++cd->aec_reference_frame_index;
-
-			if (cd->aec_reference_frame_index == cd->num_frames) {
-				GoogleRtcAudioProcessingAnalyzeRender_int16(cd->state,
-									    cd->aec_reference_buffer);
-				cd->aec_reference_frame_index = 0;
-			}
+	src_stream = sources[cd->raw_microphone_source];
+	ref_stream = sources[cd->aec_reference_source];
+	dst_stream = sinks[0];
+
+	num_of_bytes_to_process = cd->num_frames * source_get_frame_bytes(ref_stream);
+	ret = source_get_data(ref_stream, num_of_bytes_to_process, (const void **)&ref,
+			      (const void **)&ref_buf_start, &ref_buf_size);
+
+	/* problems here are extremely unlikely, as it has been checked that
+	 * the buffer contains enough data
+	 */
+	assert(!ret);
+	ref_buf_end = ref_buf_start + ref_buf_size;
+
+
+	/* can't use source_get_data_frames_available as number of available data may have changed
+	 * other processes may put some data to the buffer
+	 */
+	num_samples_remaining = num_of_bytes_to_process * source_get_channels(ref_stream) /
+				source_get_frame_bytes(ref_stream);
+
+	/* de-interlace ref buffer, convert it to float */
+	for (int i = 0; i < cd->num_frames; i++) {
+		for (channel = 0; channel < cd->num_aec_reference_channels; ++channel) {
+			cd->aec_reference_buffer_ptrs[channel][i] =
+					convert_uint16_to_float_hifi(ref[channel]);
 		}
-		num_samples_remaining -= n;
-		ref = audio_stream_wrap(ref_stream, ref);
+		ref += cd->num_aec_reference_channels;
+		if ((void *)ref >= (void *)ref_buf_end)
+			ref = (void *)ref_buf_start;
```

Comment:
I guess this won't be a lot, but we could optimise this a bit: remove this `if` in the loop and calculate first till the potential end of the buffer, and then wrap, or use an encompassing loop as in other modules. Same for source and sink below

Reply:
from my test before, wrap would be faster than we check in the for loop.